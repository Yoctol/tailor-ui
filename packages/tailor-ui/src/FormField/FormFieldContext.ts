import { createContext } from 'react';

export type Value = number | string | object | boolean | null;

const FormFieldContext = createContext<{
  invalid: boolean;
  labelId?: string;
  setValue: (value: Value) => void;
  setLabelId: (id: string) => void;
}>({
  invalid: false,
  labelId: undefined,
  setValue: () => {},
  setLabelId: () => {},
});

export default FormFieldContext;
