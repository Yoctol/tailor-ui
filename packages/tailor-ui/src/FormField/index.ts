export { default } from './FormField';
export * from './FormField';
