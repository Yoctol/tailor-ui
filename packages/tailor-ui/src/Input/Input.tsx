import React, {
  FunctionComponent,
  InputHTMLAttributes,
  KeyboardEvent,
  KeyboardEventHandler,
  ReactNode,
  forwardRef,
  isValidElement,
  useEffect,
} from 'react';
import styled, { css } from 'styled-components';
import {
  SpaceProps,
  TextAlignProps,
  WidthProps,
  space,
  textAlign,
  width,
} from 'styled-system';

import { Omit } from '../utils/type';
import { StyledButton } from '../Button';

export type Size = 'sm' | 'md' | 'lg';

export type StyledInputProps = WidthProps &
  SpaceProps &
  TextAlignProps & {
    size?: Size;
  };

export const inputStyles = css<StyledInputProps>`
  display: block;
  width: 100%;
  max-width: 100%;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray400};
  outline: none;
  background-color: ${p => p.theme.colors.light};
  color: ${p => p.theme.colors.gray700};
  line-height: ${p => p.theme.lineHeight};
  appearance: none;

  &:hover {
    border-color: ${p => p.theme.colors.primary};
  }

  &:focus {
    border-color: ${p => p.theme.colors.primary};
    box-shadow: inset 0 0 0 2px ${p => p.theme.colors.surface};
  }

  &:disabled,
  [disabled] {
    border-color: ${p => p.theme.colors.gray400};
    background-color: ${p => p.theme.colors.gray300};
    color: ${p => p.theme.colors.gray500};
    cursor: not-allowed;
  }

  &::placeholder {
    color: ${p => p.theme.colors.gray400};
  }

  ${({ size = 'md', theme: { paddings, heights, fontSizes } }) => {
    const inputSizeStyles = {
      sm: css`
        height: ${heights.sm};
        padding: 1px ${paddings.xs};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 4px ${paddings.sm};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 6px ${paddings.md};
        font-size: ${fontSizes.lg};
      `,
    };

    return inputSizeStyles[size];
  }};

  ${p => p.theme.transition /* sc-declaration */};

  ${width};
  ${space};
  ${textAlign};
`;

export const StyledInput = styled.input`
  ${inputStyles};
`;

const InputLabel = styled.span`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 0 ${p => p.theme.space[2]};
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray300};
  background-color: ${p => p.theme.colors.gray200};
  font-size: ${p => p.theme.fontSizes.base};
  line-height: 1;
`;

interface InputLabel {
  prefix?: any;
  suffix?: any;
}

const InputWrapper = styled.div<InputLabel>`
  display: flex;

  ${p =>
    p.prefix &&
    css`
      ${StyledInput /* sc-selector */} {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }

      ${InputLabel /* sc-selector */}:first-child {
        border-right: none;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }
    `};

  ${p =>
    p.suffix &&
    css`
      ${StyledInput /* sc-selector */} {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }

      ${InputLabel /* sc-selector */}:last-child {
        border-left: none;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }

      ${StyledButton /* sc-selector */} {
        flex: none;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }
    `};

  ${space};
  ${width};
`;

export type InputProps = Omit<InputHTMLAttributes<HTMLInputElement>, 'size'> & {
  /**
   * Auto select value of the input if true
   */
  autoSelect?: boolean;
  /**
   * The size of the input box
   */
  size?: Size;
  /**
   * The callback function that is triggered when Enter key is pressed
   */
  onPressEnter?: KeyboardEventHandler<HTMLInputElement>;
  /**
   * The label text displayed before (on the right side of) the input field.
   */
  prefix?: ReactNode;
  /**
   * The label text displayed after (on the right side of) the input field.
   */
  suffix?: ReactNode;
  onKeyPress?: KeyboardEventHandler<HTMLInputElement>;
  required?: boolean;
};

const Input: FunctionComponent<InputProps> = forwardRef<
  HTMLInputElement,
  InputProps
>(
  (
    {
      prefix,
      suffix,
      size = 'md',
      onPressEnter,
      onKeyPress,
      autoSelect,
      autoFocus,
      ...props
    },
    ref
  ) => {
    useEffect(() => {
      if (autoSelect && ref) {
        (ref as any).select();
      }
    }, [autoSelect, ref]);

    const handleKeyPress = (event: KeyboardEvent<HTMLInputElement>) => {
      if (onKeyPress) {
        onKeyPress(event);
      }

      if (onPressEnter && event.key === 'Enter') {
        onPressEnter(event);
      }
    };

    const input = (
      <StyledInput
        ref={ref}
        size={size as any}
        onKeyPress={handleKeyPress}
        autoFocus={autoFocus || autoSelect}
        {...props}
      />
    );

    if (prefix || suffix) {
      return (
        <InputWrapper prefix={prefix} suffix={suffix}>
          {prefix && <InputLabel>{prefix}</InputLabel>}
          {input}
          {suffix &&
            (isValidElement(suffix) ? (
              suffix
            ) : (
              <InputLabel>{suffix}</InputLabel>
            ))}
        </InputWrapper>
      );
    }

    return input;
  }
);

export default Input;
