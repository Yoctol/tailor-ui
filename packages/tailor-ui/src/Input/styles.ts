import styled, { css } from 'styled-components';
import {
  SpaceProps,
  TextAlignProps,
  WidthProps,
  space,
  textAlign,
  width,
} from 'styled-system';

import { StyledButton } from '../Button';

export type Size = 'sm' | 'md' | 'lg';

export type StyledInputProps = WidthProps &
  SpaceProps &
  TextAlignProps & {
    invalid?: boolean;
    size?: Size;
  };

export const inputStyles = css<StyledInputProps>`
  display: block;
  width: 100%;
  max-width: 100%;
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray400};
  outline: none;
  background-color: ${p => p.theme.colors.light};
  color: ${p => p.theme.colors.gray700};
  line-height: ${p => p.theme.lineHeight};
  appearance: none;

  &:hover {
    border-color: ${p => p.theme.colors.primary};
  }

  &:focus {
    border-color: ${p => p.theme.colors.primary};
    box-shadow: inset 0 0 0 2px ${p => p.theme.colors.surface};
  }

  &:disabled,
  [disabled] {
    border-color: ${p => p.theme.colors.gray400};
    background-color: ${p => p.theme.colors.gray300};
    color: ${p => p.theme.colors.gray500};
    cursor: not-allowed;
  }

  &::placeholder {
    color: ${p => p.theme.colors.gray400};
  }

  ${p =>
    p.invalid &&
    css`
      border-color: ${p.theme.colors.danger} !important;
      box-shadow: none !important;
    `}

  ${({ size = 'md', theme: { paddings, heights, fontSizes } }) => {
    const inputSizeStyles = {
      sm: css`
        height: ${heights.sm};
        padding: 1px ${paddings.xs};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 4px ${paddings.sm};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 6px ${paddings.md};
        font-size: ${fontSizes.lg};
      `,
    };

    return inputSizeStyles[size];
  }};

  ${p => p.theme.transition /* sc-declaration */};

  ${width};
  ${space};
  ${textAlign};
`;

export const StyledInput = styled.input`
  ${inputStyles};
`;

export const InputLabel = styled.span`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: 0 ${p => p.theme.space[2]};
  border: ${p => p.theme.borders.base};
  border-radius: ${p => p.theme.radii.base};
  border-color: ${p => p.theme.colors.gray300};
  background-color: ${p => p.theme.colors.gray200};
  font-size: ${p => p.theme.fontSizes.base};
  line-height: 1;
`;

interface InputWrapperProps {
  prefix?: any;
  suffix?: any;
}

export const InputWrapper = styled.div<InputWrapperProps>`
  display: flex;

  ${p =>
    p.prefix &&
    css`
      ${StyledInput /* sc-selector */} {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }

      ${InputLabel /* sc-selector */}:first-child {
        border-right: none;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }
    `};

  ${p =>
    p.suffix &&
    css`
      ${StyledInput /* sc-selector */} {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
      }

      ${InputLabel /* sc-selector */}:last-child {
        border-left: none;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }

      ${StyledButton /* sc-selector */} {
        flex: none;
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
      }
    `};

  ${space};
  ${width};
`;
