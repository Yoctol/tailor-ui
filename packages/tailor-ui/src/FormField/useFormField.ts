import { useContext, useEffect } from 'react';

import { tuplify } from '@tailor-ui/utils';

import FormFieldContext from './FormFieldContext';

const useFormField = ({
  id,
  value,
  defaultValue,
}: {
  id?: string;
  value: any;
  defaultValue: any;
}) => {
  const { invalid, setValue, setLabelId, labelId } = useContext(
    FormFieldContext
  );

  useEffect(() => {
    if (id) {
      setLabelId(id);
    }
  }, [id, setLabelId]);

  useEffect(() => {
    setValue(value || defaultValue);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return tuplify(invalid, id || labelId, setValue);
};

export { useFormField };
