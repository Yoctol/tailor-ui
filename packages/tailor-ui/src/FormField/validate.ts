import { ReactNode } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Schema } from 'yup';

import { Value } from './FormFieldContext';

export type Validator =
  | ((value?: Value) => string | null)
  | Schema<Value>
  | ({
      rule: (value?: Value) => boolean;
      message: string;
    })[];

export interface Validate {
  value?: Value;
  validationMessage?: ReactNode;
  validator?: Validator;
}

const isSchema = (obj: any): obj is Schema<Value> => obj && obj.__isYupSchema__;

export const validate = async ({
  value,
  validator,
  validationMessage,
}: Validate) => {
  if (validationMessage) {
    return {
      invalid: true,
      message: validationMessage,
    };
  }

  if (!validator) {
    return {
      invalid: false,
      message: null,
    };
  }

  if (isSchema(validator)) {
    try {
      await validator.validate(value);

      return {
        invalid: false,
        message: null,
      };
    } catch (err) {
      return {
        invalid: true,
        message: err.name === 'ValidationError' ? err.errors[0] : null,
      };
    }
  }

  if (validator instanceof Function) {
    const message = validator(value);
    return {
      invalid: Boolean(message),
      message,
    };
  }

  if (Array.isArray(validator)) {
    let returnMessage = null;
    let invalid = false;

    validator.some(({ rule, message }) => {
      if (rule(value)) {
        returnMessage = message;
        invalid = true;
        return true;
      }
      return false;
    });

    return {
      invalid,
      message: returnMessage,
    };
  }

  return {
    invalid: false,
    message: null,
  };
};
