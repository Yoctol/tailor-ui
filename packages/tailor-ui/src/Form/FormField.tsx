import React, { FunctionComponent } from 'react';
import styled, { css } from 'styled-components';

import tag from '../utils/CleanTag';
import { StyledInput } from '../Input';
import { StyledTextarea } from '../Input/Textarea';

import Hint from './Hint';

export interface FormFieldProps {
  success?: boolean;
  warning?: boolean;
  error?: boolean;
}

const StyledFormField = styled(tag.div)<FormFieldProps>`
  position: relative;
  margin-bottom: ${p => p.theme.space[2]};

  ${({ success }) =>
    success &&
    css`
      & ${StyledInput /* sc-selector */} {
        border-color: ${p => p.theme.colors.success} !important;
        box-shadow: none !important;
      }

      /* stylelint-disable-next-line */
      & ${StyledTextarea /* sc-selector */} {
        border-color: ${p => p.theme.colors.success} !important;
        box-shadow: none !important;
      }

      /* stylelint-disable-next-line */
      & ${Hint /* sc-selector */} {
        color: ${p => p.theme.colors.success};
      }
    `};

  ${({ warning }) =>
    warning &&
    css`
      & ${StyledInput /* sc-selector */} {
        border-color: ${p => p.theme.colors.warning} !important;
        box-shadow: none !important;
      }

      /* stylelint-disable-next-line */
      & ${StyledTextarea /* sc-selector */} {
        border-color: ${p => p.theme.colors.warning} !important;
        box-shadow: none !important;
      }

      /* stylelint-disable-next-line */
      & ${Hint /* sc-selector */} {
        color: ${p => p.theme.colors.warning};
      }
    `};

  ${({ error }) =>
    error &&
    css`
      & ${StyledInput /* sc-selector */} {
        border-color: ${p => p.theme.colors.danger} !important;
        box-shadow: none !important;
      }

      /* stylelint-disable-next-line */
      & ${StyledTextarea /* sc-selector */} {
        border-color: ${p => p.theme.colors.danger} !important;
        box-shadow: none !important;
      }

      /* stylelint-disable-next-line */
      & ${Hint /* sc-selector */} {
        color: ${p => p.theme.colors.danger};
      }
    `};
`;

const FormField: FunctionComponent<FormFieldProps> = props => (
  <StyledFormField {...props} />
);

export default FormField;
