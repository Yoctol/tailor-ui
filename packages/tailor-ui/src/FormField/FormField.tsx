import React, {
  FunctionComponent,
  ReactNode,
  useCallback,
  useEffect,
  useState,
} from 'react';

import { Box } from '../Layout';
import { useUID } from '../UIProvider/UIDContext';

import FormFieldContext, { Value } from './FormFieldContext';
import { Label, ValidationMessage } from './styles';
import { Validator, validate } from './validate';

export interface FormFieldProps {
  label?: string;
  required?: boolean;
  validator?: Validator;
  validationMessage?: ReactNode;
}

const FormField: FunctionComponent<FormFieldProps> = ({
  label,
  required = false,
  validator,
  validationMessage,
  children,
  ...props
}) => {
  const getUID = useUID();
  const [labelId, setLabelId] = useState(() => getUID());
  const [value, setValue] = useState<Value>();
  const [{ invalid, message }, setValidationResult] = useState({
    invalid: false,
    message: null,
  });

  const handleValidation = useCallback(async () => {
    if (value !== undefined) {
      const result = await validate({
        value,
        validator,
        validationMessage,
      });

      setValidationResult(result);
    }
  }, [validationMessage, validator, value]);

  useEffect(() => {
    handleValidation();
  }, [handleValidation]);

  return (
    <FormFieldContext.Provider
      value={{ invalid, setValue, setLabelId, labelId }}
    >
      <Box mb="16px" {...props}>
        {label && (
          <Label required={required} htmlFor={labelId}>
            {label}
          </Label>
        )}

        {children}
        {invalid && <ValidationMessage>{message}</ValidationMessage>}
      </Box>
    </FormFieldContext.Provider>
  );
};

export { FormField };
