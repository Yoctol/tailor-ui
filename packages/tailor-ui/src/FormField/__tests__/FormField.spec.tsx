import React, { useState } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { string } from 'yup';

import { fireEvent, render, wait } from 'test/test-utils';

import { FormField } from '../FormField';
import { Input } from '../../Input';

describe('FormField', () => {
  it('should render correctly', () => {
    const { container, getByLabelText } = render(
      <FormField label="Input">
        <Input id="input-id" placeholder="Placeholder" defaultValue="" />
      </FormField>
    );

    expect(getByLabelText('Input')).toBeInTheDocument();
    expect(container.querySelector('#input-id')).toBeInTheDocument();
    expect(container.querySelector('label[for=input-id]')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render generated id correctly', () => {
    const { container, getByLabelText } = render(
      <FormField label="Input">
        <Input placeholder="Placeholder" defaultValue="" />
      </FormField>
    );

    expect(getByLabelText('Input')).toBeInTheDocument();
    expect(container.querySelector('#tailor_uid_1')).toBeInTheDocument();
    expect(
      container.querySelector('label[for=tailor_uid_1]')
    ).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('validationMessage', () => {
    it('should render FormField with error message correctly', async () => {
      const { container, findByText } = render(
        <FormField label="Input" validationMessage="Error Message">
          <Input placeholder="Placeholder" defaultValue="" />
        </FormField>
      );

      const message = await findByText('Error Message');

      await wait(() =>
        expect(message.parentElement).toHaveStyle('opacity: 1; height: 22px')
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      const TextInput = () => {
        const [value, setValue] = useState('');

        return (
          <FormField
            label="Input"
            validationMessage={value === 'error' ? 'Error Message' : null}
          >
            <Input
              data-testid="input"
              value={value}
              onChange={e => setValue(e.target.value)}
            />
          </FormField>
        );
      };

      const { getByTestId, findByText, queryByText } = render(<TextInput />);

      expect(queryByText('Error Message')).not.toBeInTheDocument();

      fireEvent.change(getByTestId('input'), {
        target: {
          value: 'error',
        },
      });

      const message = await findByText('Error Message');

      expect(message).toBeInTheDocument();
    });
  });

  describe('yup schema validator', () => {
    it('should render error message when input is invalid value', async () => {
      const { findByText } = render(
        <FormField
          label="Input"
          validator={string().test(
            'is-error',
            'Error Message',
            value => value !== 'error'
          )}
        >
          <Input defaultValue="error" placeholder="Placeholder" />
        </FormField>
      );

      const message = await findByText('Error Message');

      expect(message).toBeInTheDocument();
    });

    it('should not render error message when input is valid value', () => {
      const { queryByText } = render(
        <FormField
          label="Input"
          validator={string().test(
            'is-error',
            'Error Message',
            value => value !== 'error'
          )}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(queryByText('Error Message')).not.toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      const { getByTestId, findByText, queryByText } = render(
        <FormField
          label="Input"
          validator={string().test(
            'is-error',
            'Error Message',
            value => value !== 'error'
          )}
        >
          <Input
            data-testid="input"
            defaultValue="valid input"
            placeholder="Placeholder"
          />
        </FormField>
      );

      expect(queryByText('Error Message')).not.toBeInTheDocument();
      fireEvent.change(getByTestId('input'), {
        target: {
          value: 'error',
        },
      });

      const message = await findByText('Error Message');
      expect(message).toBeInTheDocument();
    });
  });

  describe('function validator', () => {
    it('should render error message when input is invalid value', async () => {
      const { findByText } = render(
        <FormField
          label="Input"
          validator={value => (value === 'error' ? 'Error Message' : null)}
        >
          <Input defaultValue="error" placeholder="Placeholder" />
        </FormField>
      );

      const message = await findByText('Error Message');

      expect(message).toBeInTheDocument();
    });

    it('should not render error message when input is valid value', () => {
      const { queryByText } = render(
        <FormField
          label="Input"
          validator={value => (value === 'error' ? 'Error Message' : null)}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(queryByText('Error Message')).not.toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      const { getByTestId, findByText, queryByText } = render(
        <FormField
          label="Input"
          validator={value => (value === 'error' ? 'Error Message' : null)}
        >
          <Input
            data-testid="input"
            defaultValue="valid input"
            placeholder="Placeholder"
          />
        </FormField>
      );

      expect(queryByText('Error Message')).not.toBeInTheDocument();
      fireEvent.change(getByTestId('input'), {
        target: {
          value: 'error',
        },
      });

      const message = await findByText('Error Message');

      expect(message).toBeInTheDocument();
    });
  });

  describe('object validator', () => {
    it('should render error message when input is invalid value', async () => {
      const { findByText } = render(
        <FormField
          label="Input"
          validator={{
            rule: value => value === 'error1',
            message: 'error message',
          }}
        >
          <Input defaultValue="error1" placeholder="Placeholder" />
        </FormField>
      );

      const message = await findByText('error message');
      expect(message).toBeInTheDocument();
    });

    it('should render error message when input is invalid value', async () => {
      const { findByText } = render(
        <FormField
          label="Input"
          validator={[
            {
              rule: value => value === 'error1',
              message: 'error message',
            },
            {
              rule: value => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="error1" placeholder="Placeholder" />
        </FormField>
      );

      const message = await findByText('error message');
      expect(message).toBeInTheDocument();
    });

    it('should render error message when input is another invalid value', async () => {
      const { findByText } = render(
        <FormField
          label="Input"
          validator={[
            {
              rule: value => value === 'error1',
              message: 'error message',
            },
            {
              rule: value => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="error2" placeholder="Placeholder" />
        </FormField>
      );

      const message = await findByText('another error message');
      expect(message).toBeInTheDocument();
    });

    it('should not render error message when input is valid', () => {
      const { queryByText } = render(
        <FormField
          label="Input"
          validator={[
            {
              rule: value => value === 'error1',
              message: 'error message',
            },
            {
              rule: value => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input defaultValue="valid input" placeholder="Placeholder" />
        </FormField>
      );

      expect(queryByText('error message')).not.toBeInTheDocument();
      expect(queryByText('another error message')).not.toBeInTheDocument();
    });

    it('should render FormField with error message correctly when change input value', async () => {
      const { getByTestId, findByText, queryByText } = render(
        <FormField
          label="Input"
          validator={[
            {
              rule: value => value === 'error1',
              message: 'error message',
            },
            {
              rule: value => value === 'error2',
              message: 'another error message',
            },
          ]}
        >
          <Input
            data-testid="input"
            defaultValue="valid input"
            placeholder="Placeholder"
          />
        </FormField>
      );

      expect(queryByText('Error Message')).not.toBeInTheDocument();

      fireEvent.change(getByTestId('input'), {
        target: {
          value: 'error1',
        },
      });

      const message = await findByText('error message');
      expect(message).toBeInTheDocument();

      fireEvent.change(getByTestId('input'), {
        target: {
          value: 'error2',
        },
      });

      const message2 = await findByText('another error message');
      expect(message2).toBeInTheDocument();
    });
  });
});
