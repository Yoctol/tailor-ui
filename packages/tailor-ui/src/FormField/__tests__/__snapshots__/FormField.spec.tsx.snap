// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField should render correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c1 {
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: #373737;
  font-size: 1rem;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

.c2 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover {
  border-color: #3c5ad0;
}

.c2:focus {
  border-color: #3c5ad0;
  box-shadow: inset 0 0 0 2px #e8ecf5;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #bfbfbf;
  background-color: #e0e0e0;
  color: #8b8b8b;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c2::-moz-placeholder {
  color: #bfbfbf;
}

.c2:-ms-input-placeholder {
  color: #bfbfbf;
}

.c2::placeholder {
  color: #bfbfbf;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="input-id"
  >
    Input
  </label>
  <input
    class="c2"
    id="input-id"
    placeholder="Placeholder"
    value=""
  />
</div>
`;

exports[`FormField should render generated id correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c1 {
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: #373737;
  font-size: 1rem;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

.c2 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover {
  border-color: #3c5ad0;
}

.c2:focus {
  border-color: #3c5ad0;
  box-shadow: inset 0 0 0 2px #e8ecf5;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #bfbfbf;
  background-color: #e0e0e0;
  color: #8b8b8b;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c2::-moz-placeholder {
  color: #bfbfbf;
}

.c2:-ms-input-placeholder {
  color: #bfbfbf;
}

.c2::placeholder {
  color: #bfbfbf;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="tailor_uid_1"
  >
    Input
  </label>
  <input
    class="c2"
    id="tailor_uid_1"
    placeholder="Placeholder"
    value=""
  />
</div>
`;

exports[`FormField validationMessage should render FormField with error message correctly 1`] = `
.c0 {
  box-sizing: border-box;
  margin-bottom: 16px;
}

.c1 {
  display: block;
  margin-bottom: 2px;
  padding-left: 2px;
  color: #373737;
  font-size: 1rem;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
}

.c3 {
  margin-top: 2px;
  padding-left: 2px;
  color: #fc5475;
  font-size: 0.875rem;
}

.c2 {
  display: block;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border-color: #fc5475 !important;
  box-shadow: none !important;
  height: 2.25rem;
  padding: 4px 0.75rem;
  font-size: 1rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c2:hover {
  border-color: #3c5ad0;
}

.c2:focus {
  border-color: #3c5ad0;
  box-shadow: inset 0 0 0 2px #e8ecf5;
}

.c2:disabled,
.c2 [disabled] {
  border-color: #bfbfbf;
  background-color: #e0e0e0;
  color: #8b8b8b;
  cursor: not-allowed;
}

.c2::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c2::-moz-placeholder {
  color: #bfbfbf;
}

.c2:-ms-input-placeholder {
  color: #bfbfbf;
}

.c2::placeholder {
  color: #bfbfbf;
}

<div
  class="c0"
>
  <label
    class="c1"
    for="tailor_uid_1"
  >
    Input
  </label>
  <input
    class="c2"
    id="tailor_uid_1"
    placeholder="Placeholder"
    value=""
  />
  <div
    class="c3"
  >
    Error Message
  </div>
</div>
`;
